syntax = "proto3";

package replication;

service NodeService {
    rpc ExecuteTransaction(Transaction) returns (TransactionResponse) {}
    rpc PropagateUpdate(UpdateNotification) returns (AckResponse) {}
    rpc SyncUpdates(UpdateGroup) returns (AckResponse) {}
    rpc NotifyLayerSync(LayerSyncNotification) returns (AckResponse) {}
    rpc GetNodeStatus(Empty) returns (NodeStatus) {}
}

message Empty {}

message DataItem {
    int32 key = 1;
    int32 value = 2;
    int32 version = 3;
    int64 timestamp = 4;
}

message Transaction {
    enum Type {
        READ_ONLY = 0;
        UPDATE = 1;
    }
    Type type = 1;
    int32 target_layer = 2;
    repeated Operation operations = 3;
}

message Operation {
    oneof operation {
        WriteOperation write = 1;
        ReadOperation read = 2;
    }
}

message WriteOperation {
    int32 key = 1;
    int32 value = 2;
}

message ReadOperation {
    int32 key = 1;
}

message UpdateNotification {
    DataItem data = 1;
    string source_node = 2;
}

message UpdateGroup {
    repeated DataItem updates = 1;
    string source_node = 2;
    int32 layer = 3;
    int32 update_count = 4;
}

message LayerSyncNotification {
    int32 source_layer = 1;
    int32 target_layer = 2;
    repeated DataItem updates = 3;
    int32 update_count = 4;
    int64 sync_timestamp = 5;
    string source_node = 6;
}

message AckResponse {
    bool success = 1;
    string message = 2;
}

message TransactionResponse {
    bool success = 1;
    repeated DataItem results = 2;
    string error_message = 3;
}

message NodeStatus {
    string node_id = 1;
    int32 layer = 2;
    int32 update_count = 3;
    repeated DataItem current_data = 4;
}

message UpdateRequest {
    DataItem update = 1;
}